tests/Interface/core/ImageContainerTest.cpp

```diff
@@ -88,15 +93,43 @@ int main() {
   //===--------------------------------------------------------------------===//
   // Test overloading bracket operator.
   //===--------------------------------------------------------------------===//
-  Img<float, 2> testBracketOperator1(image);
+  Img<float, 2> testBracketOperator1(grayimage);
   // CHECK: 240.0
   fprintf(stderr, "%f\n", testBracketOperator1[15]);
   testBracketOperator1[15] = 90.0;
   // CHECK: 90.0
   fprintf(stderr, "%f\n", testBracketOperator1[15]);
-  const Img<float, 2> testBracketOperator2(image);
+  const Img<float, 2> testBracketOperator2(grayimage);
   // CHECK: 240.0
   fprintf(stderr, "%f\n", testBracketOperator2[15]);
 
+  //===--------------------------------------------------------------------===//
+  // Test image channels layout of RGB images from ImgContainer.
+  //===--------------------------------------------------------------------===//
+  // The test image is a RGB image with size 577 * 700.
+  // The test running directory is in <build dir>/tests/Interface/core, so the
+  // `imread` function uses the following relative path.
+
+  cv::Mat RGBimage =
+      cv::imread("../../../../tests/Interface/core/TestRGBImage.jpg");
+
+  // Represent NHWC layout by default.
+  Img<float, 4> testRGBImageLayout1(RGBimage);
+  // CHECK: 1, 577, 700, 3
+  fprintf(stderr, "%ld, %ld, %ld, %ld\n", testRGBImageLayout1.getSizes()[0],
+          testRGBImageLayout1.getSizes()[1], testRGBImageLayout1.getSizes()[2],
+          testRGBImageLayout1.getSizes()[3]);
+
+  // Represent NCHW layout with sizes = {1, 3, 577, 700}
+  intptr_t sizesInput2[4] = {1, 3, RGBimage.rows, RGBimage.cols};
+  Img<float, 4> testRGBImageLayout2(RGBimage, sizesInput2);
+  // CHECK: 1211700, 403900, 700, 1
+  fprintf(stderr, "%ld, %ld, %ld, %ld\n", testRGBImageLayout2.getStrides()[0],
+          testRGBImageLayout2.getStrides()[1],
+          testRGBImageLayout2.getStrides()[2],
+          testRGBImageLayout2.getStrides()[3]);
+
```

Comment:
Thanks for adding this test. 
I'd appreciate a small test (perhaps handwritten) for covering [this case](https://github.com/xlinsist/buddy-mlir/blob/2b93d236536f332e48e229a51da3efb0265f8f00/lib/Interface/core/ImageContainer.cpp#L53) as well, for completeness.

Reply:
It will be good but I haven't found methods to test assert statements based on original ImageContainerTest. I think the assertations so far are clear enough and the checking can be done in another commit for test updates.

Reply:
Did you search for something similar to [this](https://github.com/llvm/llvm-project/blob/e31d27e46048ccc3294d6b215dc778b3390e7834/mlir/test/Dialect/Arith/invalid.mlir#L4) directive?

Reply:
Such checking for invalid cases 

> I get the following warnings when I try to build `correlation2D` example:
> 
> ```
> ../lib/Interface/core/ImageContainer.cpp: In member function ‘void Img<T, N>::loadImg(cv::Mat, bool) [with T = float; long unsigned int N = 2]’:
> ../lib/Interface/core/ImageContainer.cpp:81:52: warning: array subscript 2 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    81 |     if (this->sizes[1] == image.rows && this->sizes[2] == image.cols &&
>       |                                         ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:82:20: warning: array subscript 3 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    82 |         this->sizes[3] == 3) {
>       |         ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:98:25: warning: array subscript 2 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    98 |     else if (this->sizes[2] == image.rows && this->sizes[3] == image.cols &&
>       |              ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:98:57: warning: array subscript 3 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    98 |     else if (this->sizes[2] == image.rows && this->sizes[3] == image.cols &&
>       |                                              ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp: In function ‘bool testImplementation(int, char**, std::ptrdiff_t, std::ptrdiff_t, std::ptrdiff_t)’:
> ../lib/Interface/core/ImageContainer.cpp:49:18: warning: array subscript 2 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    49 |       this->sizes[2] = image.cols;
>       |       ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:50:18: warning: array subscript 3 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    50 |       this->sizes[3] = 3;
> ```
> 
> Ideally, we should not trigger additional warnings in the project. Hence, can you please take care of them. Rest LGTM. Good effort!

This array-bound warning is caused by compiler not knowing layout array subscript 2 or 3 will never be reached when processing gray images with N equal to 2. So I explictly replace 

Suppress array-bounds warning since layout array subscript 2 and 3 will never
// be reached when processing gray images with N equal to 2.

> I get the following warnings when I try to build `correlation2D` example:
> 
> ```
> ../lib/Interface/core/ImageContainer.cpp: In member function ‘void Img<T, N>::loadImg(cv::Mat, bool) [with T = float; long unsigned int N = 2]’:
> ../lib/Interface/core/ImageContainer.cpp:81:52: warning: array subscript 2 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    81 |     if (this->sizes[1] == image.rows && this->sizes[2] == image.cols &&
>       |                                         ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:82:20: warning: array subscript 3 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    82 |         this->sizes[3] == 3) {
>       |         ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:98:25: warning: array subscript 2 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    98 |     else if (this->sizes[2] == image.rows && this->sizes[3] == image.cols &&
>       |              ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:98:57: warning: array subscript 3 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    98 |     else if (this->sizes[2] == image.rows && this->sizes[3] == image.cols &&
>       |                                              ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp: In function ‘bool testImplementation(int, char**, std::ptrdiff_t, std::ptrdiff_t, std::ptrdiff_t)’:
> ../lib/Interface/core/ImageContainer.cpp:49:18: warning: array subscript 2 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    49 |       this->sizes[2] = image.cols;
>       |       ~~~~~~~~~~~^
> ../lib/Interface/core/ImageContainer.cpp:50:18: warning: array subscript 3 is above array bounds of ‘intptr_t [2]’ {aka ‘long int [2]’} [-Warray-bounds]
>    50 |       this->sizes[3] = 3;
> ```
> 
> Ideally, we should not trigger additional warnings in the project. Hence, can you please take care of them. Rest LGTM. Good effort!

This array-bound warning is caused by compiler not knowing layout array subscript 2 and 3 will never be reached when processing gray images with N equal to 2. So I change the if-else condition from `image.channels() == 1` to `N == 2` to erase the warning (these conditions are essentially the same under what ImageContainer supports) and add additional comments.

I present a new commit for the modification, thanks again for your review! @meshtag

Reply:
Can you create an issue to address this.